### DISCUSSION TITLE
Is AesGcm thread safe?

### DISCUSSION THREAD

**Discussion by kostya9** [dotnet/runtime#71198] [Q&A] [answered]
I would like to cache the AesGcm instance in a way that only one instance is created per key, but I have not found any documentation if I can use one AesGcm instance in multiple threads concurrently (imagine a web API use-case).

Can I reuse one AesGcm instance across multiple threads?

**Comment by PathogenDavid** [comment, marked as answer]
You cannot.

Looking at the implementation of [`EncryptCore`](https://github.com/dotnet/runtime/blob/ddc4f95432facf2deebfa85bad7143469dca0129/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesGcm.OpenSsl.cs#L30) (specifically the OpenSSL implementation) you can see there's state being stored in the `_ctxHandle` that would not appreciate differing across threads.

**Reply by kostya9** [reply]
Ah, I feared that's the case. I came to the same conclusion but wanted to double-check and get a second opinion. Thanks!

**Reply by TonyValenti** [reply]
You could likely wrap the class and put the appropriate locks in place though.

**Reply by Xyncgas** [reply]
Or use Bouncy Castle !

**Comment by GrabYourPitchforks** [comment]
@PathogenDavid answered the question for this type specifically.

But _in general_, the rule of thumb is that instance APIs are not thread-safe for multiple callers. If there are exceptions, they'll be called out in the API documentation. See [the `Dictionary<TKey, TValue>` docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-6.0#thread-safety) and [the `HttpClient` docs](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-6.0#thread-safety) for examples of what it looks like when the docs call out these special-cases.
